📖 Add new documentation explaining the usage of Pprof (#4160)
kubernetes-sigs/kubebuilder: docs/book/src/cronjob-tutorial/pprof-tutorial.md

# Using Pprof for Controller Profiling

Pprof is a Go based library designed for profiling the binary in order to gather CPU, and Memory related data. Pprof is help in finding any sort of bottle necks that are either appearing at Memory, CPU, or both levels.

> Pprof itself runs as a tiny HTTP server as part of the the Controller. You can extract the statistics, and visualize them on an interactive website. Furthermore, **you don't need to install this package seperately**. Pprof comes packaged as part of Kubebuilder `controller-runtime` [dependency](https://github.com/kubernetes-sigs/controller-runtime/pull/1943) within the scafolded project.

Reviewer: We need try to be accurate with the technical terms. 
So, it seems for me that we could improve it with such as, wdyt:

```suggestion
`pprof` is a Go profiling tool integrated within the `controller-runtime` library, designed to help identify performance bottlenecks in areas such as CPU usage, memory allocation, and more. This profiling tool, from the Go package [`net/http/pprof`][pprof-go-docs], is especially useful for diagnosing performance issues by providing detailed insights into resource usage and performance during the runtime of your application.

`pprof` is integrated into the HTTP server used by the `controller-runtime` manager, allowing you to collect profiling data via HTTP endpoints. This data can then be analyzed using tools like `go tool pprof` or visualized through command-line or web-based interfaces. This feature is built into the `controller-runtime` library, so there is no need to install it separately.

The `controller-runtime` [Manager options][manager-options-doc] provide an easy way to enable `pprof`, allowing you to gather detailed runtime metrics for improving the performance of your controllers.
```

Also, see that for the links we use alias, so that we can keep things maintained easily. The definition  of the alias should be set at the bottom of the page.
```
[pprof-go-docs]: https://pkg.go.dev/net/http/pprof
[manager-options-doc]: https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/manager
```

> For further information about the Pprof usage, check out the official [docs](https://github.com/google/pprof).

Reviewer: I would add the link alias at the bottom and all places that we say Pprof I would use the link instead of 

[Pprof][pprof-project] so users can just click on it.

# How to use Pprof in your Codebase

1. In your `cmd/main.go` file, specify the `PprofBindAddress` URL field as part of Controller Manager's configurations:

    ```golang
    mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
      ....
      Scheme:                 scheme,
      LeaderElection:         enableLeaderElection,

>       PprofBindAddress:       "127.0.0.1:8082", // Specify your own URL here.

Reviewer 2: it it not url, it is host interface on which server will listening

Author: @grzesuav what's the difference exactly? Network Interface sounds more like NIC (Network Interface Card) which the end user doesn't usually care about.

Reviewer: We need add a valid comment/term, see the definition in the lib:

https://github.com/kubernetes-sigs/controller-runtime/blob/38546806f2faf5973e3321a7bd5bb3afdbb5767d/pkg/manager/manager.go#L237-L242

So, I would suggest
```suggestion
      // PprofBindAddress is the TCP address that the controller should bind to
      // for serving pprof. Specify the manager address and the port that should be bind.
      PprofBindAddress:       "127.0.0.1:8082", 
```

Author: Sounds reasonable.

Reviewer 2: each network interface has ip address bind to it,from which to are special - like `127.0.0.1` is localhost and `0.0.0.0` means - bind to all local interfaces. I am not a network engineer so my explanation can be a little dump, but think if computer has more than one network card, each ot the network cards is identified by different ip address locally, so you can expose service on one network and not expose on the another.

> # Using Pprof for Controller Profiling

Reviewer: ```suggestion
# Monitoring Performance with Pprof
```

Maybe ?!

Reviewer: Also, we need to add the link to the reference menu:

https://github.com/kubernetes-sigs/kubebuilder/blob/master/docs/book/src/reference/reference.md

That need too be short like `Monitoring with Pprof`

After do the changes you can check the doc and visualize its final result, see: 

<img width="700" alt="Screenshot 2024-09-13 at 08 11 35" src="https://github.com/user-attachments/assets/199df341-3d77-4f89-b47b-3c49a043bc4f">

For your PR the docs are rendered based in your changes in : https://deploy-preview-4160--kubebuilder.netlify.app/

Pprof is a Go based library designed for profiling the binary in order to gather CPU, and Memory related data. Pprof is helpful in finding any sort of bottlenecks that are appearing either at Memory, CPU, or both levels.

Pprof itself runs as part of the the Controller. You can extract the statistics, and visualize them on an interactive website. Furthermore, **you don't need to install this package seperately**. Pprof comes packaged as part of Kubebuilder `controller-runtime` [dependency](https://github.com/kubernetes-sigs/controller-runtime/pull/1943) within the scafolded project.

For further information about the Pprof usage, check out the official [docs](https://github.com/google/pprof).

> 

Reviewer: We need to add a note with warning here.
Users need to be aware that they should not use it in Prod 
See that we have a css to generate notes [`<aside class="warning">`](https://github.com/search?q=repo%3Akubernetes-sigs%2Fkubebuilder+%3Caside+class%3D%22warning%22%3E&type=code)

Something such as :

```suggestion
<aside class="warning">
<h1>Pprof Not Recommended for Production</h1>

While [Pprof][link-alias] is an excellent tool for profiling and debugging, it is not recommended to leave it enabled in production environments. The primary reasons are:

- **Security Risk**: The profiling endpoints expose detailed information about your application's performance and resource usage, which could be exploited if accessed by unauthorized users.
- **Overhead**: Continuous profiling can introduce performance overhead, especially under heavy load, potentially impacting production workloads.

</aside>

```

Author: @camilamacedo86 warning note should be at the beginning or end of the document?

Reviewer: Yes that is the best place to warning the people
After the overview before someone try it out

> # How to use Pprof in your Codebase

Reviewer: It is a sub-section so need to be `##` and not `#`, either I think something more like seems fit better, wdyt

```suggestion
## Enabling Pprof 

To enable `pprof` for your manager, follow these steps:
```

Reviewer: We do not use Codebase term usually, that does not seems to fit
Need to be something like  Enabling Pprof

> 1. In your `cmd/main.go` file, specify the `PprofBindAddress` URL field as part of Controller Manager's configurations:

Reviewer: ```suggestion
1. **Update Manager Options**  

In your `cmd/main.go` file, you can enable the `pprof` endpoints by configuring the `controller-runtime` manager [options][link-alias]. Add the following code snippet to include the profiling server:
```

Reviewer: @TAM360

    ```golang
    mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
      ....

>       Scheme:                 scheme,
>       LeaderElection:         enableLeaderElection,
>       // PprofBindAddress is the TCP address that the controller should bind to

Reviewer: ```suggestion
      // PprofBindAddress is the TCP address that the controller should bind to
```

We do not need those options right?

Author: Nope. Only `PprofBindAddress` is required.

      // for serving pprof. Specify the manager address and the port that should be bind.

>       PprofBindAddress:       "127.0.0.1:8082",

Reviewer: ```suggestion
      PprofBindAddress:       ":8082", // Set the port where pprof will be available
```

I think you do not need set the localhost IP just the port to bind

Author: @camilamacedo86 `:8082` means that it listens on `localhost` or `0.0.0.0`? IMO, keeping the IP addresses would be much helpful. It's standard syntax used across all apps that involve HTTP server setup.

Author: In short, use `0.0.0.0` instead of `127.0.0.1`?

Reviewer: In short, if we use `:8082` that works for both as fair I am aware of.

Author: Gonna verify this first before pushing any changes.

Author: @camilamacedo86  `:8082` works just fine. I've updated the document.

      ...
    })
    ```

> 2. Reinstall the Controller binary.

Reviewer: The following does not seems accurate.
The `make install` will install on the Custer the CRDs: https://github.com/kubernetes-sigs/kubebuilder/blob/d26664f3fa472e98364ce25ec2cc786f25de748b/testdata/project-v4/Makefile#L141-L143

Give a look in the QuickStart, to check it out.

I think we can now here mainly only say, 

```suggestion
2. **Test It Out**

After enabling [Pprof][link-alias], you need to build and deploy your controller to test it out. Follow the steps in the [Quick Start guide][quick-start-run-it] to run your project locally or on a cluster.

[quick-start-run-it]: ./docs/book/src/quick-start.md#test-it-out
```

>     ```bash
>     make install
>     ```

Reviewer: ```suggestion

```

> 3. Start your controller on your localhost:
> 
>     ```bash
>     export ENABLE_WEBHOOKS=false # If you're not using a Webhook. Else, set it to true.
>     make run
>     ```

Reviewer: ```suggestion
```

We can remove this one. Not everybody is or not using webhook-hooks either the explanation how to run is the docs. By last who will enable it probably is not running the project by the first time.

4. Deploy your custom resource on Kind, or any other Kubernetes cluster. Wait for the pods to be ready.

    ```bash
    kubectl apply -f batch_v1_cronjob.yaml

>     ```

Reviewer: we do not need to apply the CR/samples right.
Now, I think is only explain how to access the profile data
PS.: the path for the sample is not accurate. The samples are scaffold under config/samples/

Author: @camilamacedo86 I've added this step with the assumption that the end user hasn't created a particular CRD based resource. By not having mentioning this, the reader may get confused. Basically, all the steps that I've are the ones that i ran as a beginner.

> 5. Use `curl` to export the profiling statistics into a file. Note that the **URL hit should be same as the one specified in the `PprofBindAddress` field of your Controller's Manager**.
> 
>     ```bash
>     curl -s "http://127.0.0.1:8082/debug/pprof/profile" > ./cpu-profile.out
>     ```
> 
> 6. Visualize the file on an interactive dashboard.
> 
>     ```bash
>     # Go tool will open a session on port 8080.
>     # You can change this as per your own need.
>     go tool pprof -http=:8080 ./cpu-profile.out
>     ```
> 
>     Visualizaion resutl will vary depending on the deployed workload, and the Controller's behavior.
>     However, you'll see a dashboard on your browser similar to this one:
> 
>     ![pprof-result-visualization](./images/pprof-result-visualization.png)

Reviewer: I would not add the image because that can change then is a suck. 
Or docs get outdated. 

I think here we would have something more like

```suggestion
3. **Access Pprof Data**

Once pprof is enabled and your solution is running, you can access the profiling data by navigating to the following URL:

````shell
http://localhost:<bind-port>/debug/pprof/
````
This will allow you to explore various profiling endpoints such as CPU, memory, and goroutine traces.

Furthermore, to analyse the profiling data, you can either use go tool pprof from the command line, or tools like pprof’s web-based interface. For example:

````shell
# To collect 20 seconds of CPU profile data 
go tool pprof http://localhost:<bind port>/debug/pprof/profile?seconds=20
```

Author: IMO, having a screenshot will provided more clarity to the readers about the end result. @camilamacedo86 WDYT?

Reviewer: Ok if you want to keep it that is fine

The `controller-runtime` [Manager options][manager-options-doc] provide an easy way to enable `pprof`, allowing you to gather detailed runtime metrics for improving the performance of your controllers.

For further information about the Pprof, check out the official [Github repository][github-repo].

<aside class="warning">
<h1>Pprof Not Recommended for Production</h1>

While [Pprof][github-repo] is an excellent tool for profiling and debugging, it is not recommended to leave it enabled in production environments. The primary reasons are:

1. **Security Risk**: The profiling endpoints expose detailed information about your application's performance and resource usage, which could be exploited if accessed by unauthorized users.
2. **Overhead**: Continuous profiling can introduce performance overhead, especially under heavy load, potentially impacting production workloads.

</aside>

> ## How to use Pprof in your Codebase

Reviewer: ```suggestion
## How to use Pprof?
```

> 1. In your `cmd/main.go` file, add the following `controller-runtime` manager field to enable the `pprof` endpoints:

Reviewer: ```suggestion
1. **Enabling Pprof**

In your `cmd/main.go` file, add the following `controller-runtime` manager field to enable the `pprof` endpoints:
```

<aside class="warning">
<h1>Pprof Not Recommended for Production</h1>

While [Pprof][github-repo] is an excellent tool for profiling and debugging, it is not recommended to leave it enabled in production environments. The primary reasons are:

1. **Security Risk**: The profiling endpoints expose detailed information about your application's performance and resource usage, which could be exploited if accessed by unauthorized users.
2. **Overhead**: Continuous profiling can introduce performance overhead, especially under heavy load, potentially impacting production workloads.

</aside>

## How to use Pprof in your Codebase

1. In your `cmd/main.go` file, add the following `controller-runtime` manager field to enable the `pprof` endpoints:

    ```golang
    mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
      ...
      // PprofBindAddress is the TCP address that the controller should bind to
      // for serving pprof. Specify the manager address and the port that should be bind.
      PprofBindAddress:       ":8082",
      ...
    })
    ```

2. Start your controller on your localhost:

    ```bash

>     make run

Reviewer: we have more than one to run so , we should not add this info here
Who will looking for `Monitoring Performance with Pprof` is skilled in KB already. Anyway, we can link the Quick start, see: https://github.com/kubernetes-sigs/kubebuilder/pull/4160#discussion_r1758291378

Author: > Who will looking for Monitoring Performance with Pprof s skilled in KB already.

@camilamacedo86 perhaps the end user is skilled in K8s. However, I would like to keep things more explicit.

> Anyway, we can link the Quick start,

I think there's no need to have a separate quick start guide. The tutorial itself is so small, and straightforward that it make the separate quick start guide redundant. This is how I see.

WDYT?

Reviewer: WDYT:

```
2. **Test It Out**

After enabling [Pprof][link-alias], you need to build and deploy your controller to test it out. Follow the steps in the [Quick Start guide][quick-start-run-it] to run your project locally or on a cluster.
[quick-start-run-it]: ./docs/book/src/quick-start.md#test-it-out.

Then, you can apply your CRs/samples in order to monitor the performance of its controllers.
```

Author: @camilamacedo86 looks OK.

    ```

3. Deploy your custom resource on Kind, or any other Kubernetes cluster. Wait for the pods to be ready.

    ```bash

>     kubectl apply -f config/samples/batch_v1_cronjob.yaml

Reviewer: Before apply A CR we need install the CRDs
Anyway, why we need to add that here?

not necesary will be this name or will cerate a POD and etc. 
I think the doc needs to be more generic.

Author: > not necesary will be this name or will cerate a POD and etc.
I think the doc needs to be more generic.

How about `Wait for the custom resource to be ready` statement?

Reviewer: By adding in the step 2 : https://github.com/kubernetes-sigs/kubebuilder/pull/4160/files#r1759293970 we can remove this one one.

    ```

> 4. Use `curl` to export the profiling statistics into a file. Note that the **URL hit should be same as the one specified in the `PprofBindAddress` field of your Controller's Manager**.

Reviewer: ```suggestion
3. Exporting the data

By using `curl`, we can export the profiling statistics into a file.
```

    ```bash

>     curl -s "http://127.0.0.1:8082/debug/pprof/profile" > ./cpu-profile.out

Reviewer: ```suggestion
    curl -s "http://127.0.0.1:<bind-port>/debug/pprof/profile" > ./cpu-profile.out
```

    ```

> 5. Visualize the results on an interactive dashboard.

Reviewer: is possible to really interact within?
What kind of interactions? 
Is not that static? Do you input data and have response?

Also, I am not sure if the wording dashboard is the best. Maybe diagram?

Author: @camilamacedo86 

1. `Browser` would be better choice than either. Visualization already points some sort of graphics and/or graphs.
2. Pprof dashboard provides couple of tools for interacting purposes such as full text search, flame graph, zoom in/out, metrics breakdown by function etc.

![code-level-profiling-stats](https://github.com/user-attachments/assets/872dea4e-6009-4502-863e-28f27ff66d81)
![flame-graph](https://github.com/user-attachments/assets/1112cbe3-4616-4c6b-ae15-05b10d3c14b0)
